trees_emit.h

```diff
@@ -45,13 +45,13 @@ extern Z_INTERNAL const int base_dist[D_CODES];
     uint32_t total_bits = bi_valid + len;\
     send_bits_trace(s, val, len);\
     sent_bits_add(s, len);\
-    if (total_bits < BIT_BUF_SIZE) {\
-        bi_buf |= val << bi_valid;\
-        bi_valid = total_bits;\
-    } else if (bi_valid == BIT_BUF_SIZE) {\
+    if (bi_valid >= BIT_BUF_SIZE) {\
```

Comment:
Moving this check earlier avoids need for checking range of `len` as on next iteration it will be `bi_valid` here. We can't rely on `total_bits` when using static analyser as exact value can't be determined on compile time. We need to use `>=` here due to not knowing exact value of `bi_valid` on compile time, when using static analyser it might be either less or more than `BIT_BUF_SIZE`.

Reply:
This change looks better. I think there might still be slight performance issue because the most common path is not the first if statement, but perhaps @Dead2 can do another test.

Reply:
Adding unlikely hint shows another decrease in code size, but unfortunately results in a further slowdown in my (repeated) tests.

Reply:
> Adding unlikely hint shows another decrease in code size, but unfortunately results in a further slowdown in my (repeated) tests.

Then we don't add `UNLIKELY` hint... I'm not sure we can completely avoid slowdown as even without static analyser, we need the `bi_valid == BIT_BUF_SIZE` test to be first to get valid generated assembly code.

Reply:
What about using `max(bi_valid, BIT_BUF_SIZE)`?

Reply:
> What about using `bi_buf |= val << (bi_valid > BIT_BUF_SIZE ? BIT_BUF_SIZE: bit_valid)`? The extra if statement stays in the path that needs it.

Shifting by exactly `BIT_BUF_SIZE` already overflows...